File: app/src/main/java/net/feheren_fekete/applist/MainActivity.java
package net.feheren_fekete.applist;

import android.appwidget.AppWidgetManager;
import android.content.Intent;
import android.os.Bundle;
import android.preference.PreferenceManager;

import net.feheren_fekete.applist.applistpage.ApplistPageFragment;
import net.feheren_fekete.applist.applistpage.ShortcutHelper;
import net.feheren_fekete.applist.launcher.LauncherFragment;
import net.feheren_fekete.applist.launcher.pageeditor.PageEditorFragment;
import net.feheren_fekete.applist.launcher.pagepicker.PagePickerFragment;
import net.feheren_fekete.applist.widgetpage.WidgetHelper;
import net.feheren_fekete.applist.widgetpage.WidgetPageFragment;
import net.feheren_fekete.applist.widgetpage.MyAppWidgetHost;
import net.feheren_fekete.applist.settings.SettingsUtils;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import static org.koin.java.KoinJavaComponent.get;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    public static final String ACTION_RESTART =
            MainActivity.class.getCanonicalName()+ "ACTION_RESTART";

    private ShortcutHelper mShortcutHelper = get(ShortcutHelper.class);
    private WidgetHelper mWidgetHelper = get(WidgetHelper.class);
    private SettingsUtils mSettingsUtils = get(SettingsUtils.class);

    private MyAppWidgetHost mAppWidgetHost;
    private AppWidgetManager mAppWidgetManager;
    private boolean mIsHomePressed;
    private boolean mShouldHandleIntent;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        PreferenceManager.setDefaultValues(getApplicationContext(), R.xml.preferences, false);
        mSettingsUtils.applyColorTheme(this);

        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_activity);

        mAppWidgetManager = AppWidgetManager.getInstance(getApplicationContext());
        mAppWidgetHost = new MyAppWidgetHost(getApplicationContext(), 1234567);

        showLauncherFragment(-1);
        mShouldHandleIntent = true;
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
        mShouldHandleIntent = true;
        if (Intent.ACTION_MAIN.equals(intent.getAction()) && intent.hasCategory(Intent.CATEGORY_HOME)) {
            // This occurs when the Home button is pressed.
            // Be careful! It may not be true in the future or on some devices.
            mIsHomePressed = true;
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mWidgetHelper.handleActivityResult(requestCode, resultCode, data, mAppWidgetHost);
    }

    @Override
    protected void onStart() {
        super.onStart();
        mAppWidgetHost.startListening();
    }

    @Override
    protected void onResume() {
        super.onResume();
        EventBus.getDefault().register(this);
        if (mShouldHandleIntent) {
            mShouldHandleIntent = false;
            handleIntent(getIntent());
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        EventBus.getDefault().unregister(this);
    }

    @Override
    protected void onStop() {
        super.onStop();
        mAppWidgetHost.stopListening();
    }

    @Override
    public void onBackPressed() {
        // Don't exit on back-press. We are a launcher.
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onShowPageEditorEvent(WidgetPageFragment.ShowPageEditorEvent event) {
        showPageEditorFragment();
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onShowPageEditorEvent(ApplistPageFragment.ShowPageEditorEvent event) {
        showPageEditorFragment();
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onShowPagePickerEvent(WidgetHelper.ShowPagePickerEvent event) {
        showPagePickerFragment(
                getResources().getString(R.string.page_picker_pin_widget_title),
                getResources().getString(R.string.page_picker_message),
                event.data);
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onShowPagePickerEvent(WidgetPageFragment.ShowPagePickerEvent event) {
        showPagePickerFragment(
                getResources().getString(R.string.page_picker_move_widget_title),
                getResources().getString(R.string.page_picker_message),
                event.data);
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onPageEditorDoneEvent(PageEditorFragment.DoneEvent event) {
        showLauncherFragment(-1);
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onPageEditorPageTappedEvent(PageEditorFragment.PageTappedEvent event) {
        if (mWidgetHelper.handlePagePicked(this, event.pageData, event.requestData)) {
            showLauncherFragment(event.pageData.getId());
        }
    }

    public boolean isHomePressed() {
        boolean wasHomePressed = mIsHomePressed;
        mIsHomePressed = false;
        return wasHomePressed;
    }

    public MyAppWidgetHost getAppWidgetHost() {
        return mAppWidgetHost;
    }

    public AppWidgetManager getAppWidgetManager() {
        return mAppWidgetManager;
    }

    private void handleIntent(@Nullable Intent intent) {
        boolean handled;
        if (intent != null) {
            handled = mShortcutHelper.handleIntent(this, intent);
            if (!handled) {
                handled = mWidgetHelper.handleIntent(this, intent, mAppWidgetManager, mAppWidgetHost);
            }
            if (!handled) {
                if (ACTION_RESTART.equals(intent.getAction())) {
                    Intent restartIntent = new Intent(this, MainActivity.class);
                    restartIntent.setAction(Intent.ACTION_MAIN);
                    restartIntent.addCategory(Intent.CATEGORY_HOME);
                    setIntent(restartIntent);
                    recreate();
                }
            }
        }
    }

    private void showLauncherFragment(long activePageId) {
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.main_activity_fragment_container, LauncherFragment.newInstance(activePageId))
                .commit();
    }

    private void showPageEditorFragment() {
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.main_activity_fragment_container,
                        PageEditorFragment.newInstance(true, false, new Bundle()))
                .commit();
    }

    private void showPagePickerFragment(String title,
                                        String message,
                                        Bundle data) {
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.main_activity_fragment_container,
                        PagePickerFragment.newInstance(title, message, data))
                .commit();
    }

}

File: app/src/main/java/net/feheren_fekete/applist/MainActivity.kt
package net.feheren_fekete.applist

import android.appwidget.AppWidgetHost
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.preference.PreferenceManager
import androidx.appcompat.app.AppCompatActivity
import net.feheren_fekete.applist.applistpage.ShortcutHelper
import net.feheren_fekete.applist.launcher.LauncherFragment
import net.feheren_fekete.applist.settings.SettingsUtils
import net.feheren_fekete.applist.utils.ProgressDialog
import net.feheren_fekete.applist.utils.WriteSettingsPermissionHelper
import net.feheren_fekete.applist.widgetpage.WidgetHelper
import org.koin.android.ext.android.inject


class MainActivity : AppCompatActivity() {

    private val applistLog: ApplistLog by inject()
    private val shortcutHelper: ShortcutHelper by inject()
    private val widgetHelper: WidgetHelper by inject()
    private val appWidgetHost: AppWidgetHost by inject()
    private val settingsUtils: SettingsUtils by inject()
    private val writeSettingsPermissionHelper: WriteSettingsPermissionHelper by inject()

    private var isHomePressed = false
    private var shouldHandleIntent = false
    private val handler = Handler()

    fun isHomePressed(): Boolean {
        val wasHomePressed = isHomePressed
        isHomePressed = false
        return wasHomePressed
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        PreferenceManager.setDefaultValues(applicationContext, R.xml.preferences, false)
        settingsUtils.applyColorTheme(this)

        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)

        showLauncherFragment(-1)
        shouldHandleIntent = true
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        shouldHandleIntent = true
        if (Intent.ACTION_MAIN == intent.action && intent.hasCategory(Intent.CATEGORY_HOME)) {
            // This occurs when the Home button is pressed.
            // Be careful! It may not be true in the future or on some devices.
            isHomePressed = true
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        widgetHelper.handleActivityResult(requestCode, resultCode, data)
    }

    override fun onStart() {
        super.onStart()
        appWidgetHost.startListening()
    }

    override fun onResume() {
        super.onResume()
        if (shouldHandleIntent) {
            shouldHandleIntent = false
            handleIntent(intent)
        }
        if (!writeSettingsPermissionHelper.hasWriteSettingsPermission()) {
            writeSettingsPermissionHelper.requestWriteSettingsPermission(this)
        }
    }

    override fun onStop() {
        super.onStop()
        appWidgetHost.stopListening()
    }

    override fun onBackPressed() {
        // Don't exit on back-press. We are a launcher.
    }

    private fun handleIntent(intent: Intent?) {
        var handled: Boolean
        if (intent != null) {
            handled = shortcutHelper.handleIntent(this, intent)
            if (!handled) {
                handled = widgetHelper.handleIntent(this, intent)
            }
            if (!handled) {
                // TODO: Move this "restart" code to the settings activity.
                if (ACTION_RESTART == intent.action) {
                    val progressDialog = ProgressDialog(R.string.settings_restarting_app)
                    progressDialog.show(supportFragmentManager, "ProgressDialog")
                    handler.postDelayed(::restartApp, 1000)
                }
            }
        }
    }

    private fun restartApp() {
        val packageManager = getPackageManager()
        val intent = packageManager.getLaunchIntentForPackage(getPackageName())
        if (intent == null) {
            applistLog.log(RuntimeException("Cannot restart app: null intent"))
            return
        }
        val componentName = intent.component
        if (componentName == null) {
            applistLog.log(RuntimeException("Cannot restart app: null component name"))
            return
        }
        val mainIntent = Intent.makeRestartActivityTask(componentName)
        try {
            startActivity(mainIntent)
            Runtime.getRuntime().exit(0)
        } catch (e: Exception) {
            applistLog.log(RuntimeException("Cannot restart app", e))
        }
    }

    private fun showLauncherFragment(activePageId: Long) {
        supportFragmentManager
            .beginTransaction()
            .replace(
                R.id.main_activity_fragment_container,
                LauncherFragment.newInstance(activePageId)
            )
            .commit()
    }

    companion object {

        private val TAG = MainActivity::class.java.simpleName

        val ACTION_RESTART = MainActivity::class.java.canonicalName!! + "ACTION_RESTART"
    }

}
